File: src/Other/AddArrayFormInteger.java
package Other;

import java.util.LinkedList;
import java.util.List;

// https://leetcode.com/problems/add-to-array-form-of-integer/
public class AddArrayFormInteger {
    public List<Integer> addToArrayForm(int[] num, int k) {
        LinkedList<Integer> res = new LinkedList<>();
        int carry = 0;
        int index = num.length - 1;
        while (k > 0 || index >= 0) {
            int curK = k % 10;
            int curA = index >= 0 ? num[index] : 0;
            int curDigitSum = curK + curA + carry;
            int toBeAdded = curDigitSum % 10;
            carry = curDigitSum / 10;
            index--;
            k /= 10;
            res.addFirst(toBeAdded);
        }
        if (carry != 0) {
            res.addFirst(1);
        }
        return res;
    }

    public static void main(String[] args) {
        AddArrayFormInteger a = new AddArrayFormInteger();
        int[] num = {1, 2, 0, 0};
        int k = 993400;
        System.out.println(a.addToArrayForm(num, k));
    }
}

File: py/AddArrayFormInteger.py
# https://leetcode.com/problems/add-to-array-form-of-integer/description/
# Math

from typing import List


class Solution:
    def addToArrayForm(self, num: List[int], k: int) -> List[int]:
        res, carry, n = [], 0, len(num) - 1

        # only when k > 0 makes sense to res, but num needs to be exhausted
        while k > 0 or n >= 0:
            sum = carry
            if k > 0:
                sum += k % 10
                k //= 10
            if n >= 0:
                sum += num[n]
                n -= 1
            res.append(sum % 10)
            carry = sum // 10  # carry > 0 when sum is larger than 10
        if carry > 0:
            res.append(carry)
        return res[::-1]


s = Solution()
num = [2, 1, 5]
k = 806
print(s.addToArrayForm(num, k))
